raiden_libs/network/matrix/client.py

```diff
@@ -21,31 +23,60 @@ class GMatrixHttpApi(MatrixHttpApi):
     A wrapper around MatrixHttpApi to limit the number
     of concurrent requests we make to the number of connections
     available to us in requests.Session connection pool size.
+
+    Args:
+        pool_maxsize: max size of underlying/session connection pool
+        retry_timeout: for how long should a single request be retried if it errors
+        retry_delay: delay to way between a retries inside a request.
+            If a callable, should return a float or an iterator of floats with each delay
     """
     def __init__(
             self,
             *args,
-            max_retries: int = 3,
             pool_maxsize: int = 10,
+            retry_timeout: int = 60,
+            retry_delay: Union[float, Callable[[], Union[float, Iterable[float]]]] = 1,
             **kwargs,
     ) -> None:
         super().__init__(*args, **kwargs)
 
-        http_adapter = HTTPAdapter(
-            max_retries=max_retries,
-            pool_maxsize=pool_maxsize,
-        )
-        https_adapter = HTTPAdapter(
-            max_retries=max_retries,
-            pool_maxsize=pool_maxsize,
-        )
+        http_adapter = HTTPAdapter(pool_maxsize=pool_maxsize)
+        https_adapter = HTTPAdapter(pool_maxsize=pool_maxsize)
         self.session.mount('http://', http_adapter)
         self.session.mount('https://', https_adapter)
+
         self.lock = Semaphore(pool_maxsize)
+        self.retry_timeout = retry_timeout
+        self.retry_delay = retry_delay
 
     def _send(self, *args, **kwargs):
-        with self.lock:
-            return super()._send(*args, **kwargs)
+        # we use an infinite loop + time + sleep instead of gevent.Timeout
+        # to be able to re-raise the last exception instead of declaring one beforehand
+        started = time.time()
+        if callable(self.retry_delay):
+            retry_delay = self.retry_delay()  # should return a float or an iterable
+        else:
+            retry_delay = self.retry_delay  # is a float
+        if not hasattr(retry_delay, '__iter__'):
+            retry_delay = repeat(retry_delay)  # a float is repeated
+        for delay in retry_delay:
+            try:
+                with self.lock:
```

Comment:
Please rename to `semaphore` to make it clear that this is not a single access lock.